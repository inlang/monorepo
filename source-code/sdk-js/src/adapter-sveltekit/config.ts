import { stat } from "node:fs/promises"

export const doesPathExist = async (path: string) => !!(await stat(path).catch(() => undefined))

export const getConfig = async () => {
	// TODO: read information from files
	const isSPA = false
	const isStatic = false

	const srcFolder = process.cwd() + '/src'
	const rootRoutesFolder = srcFolder + '/routes/(app)' + (isSPA ? '' : '/[lang]')

	const hasAlreadyBeenInitialized = await doesPathExist(rootRoutesFolder)

	return {
		isSPA,
		isStatic,
		srcFolder,
		rootRoutesFolder,
		hasAlreadyBeenInitialized,
	}
}