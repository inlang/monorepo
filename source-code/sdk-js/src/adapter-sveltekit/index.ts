export { vitePlugin, rollupPlugin } from "./plugin.js"
