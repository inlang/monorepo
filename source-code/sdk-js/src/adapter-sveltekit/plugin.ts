import { createUnplugin } from "unplugin"

const unplugin = createUnplugin(() => {
	return {
		name: "inlang-sveltekit-adapter",
		buildStart() {
			console.log("Here comes some inlang magic")
		},
		/* 
			this is how we could potentially transform our js files.

			Recast mentioned here, by rich harris: https://github.com/Rich-Harris/magic-string#magic-string

			async transform(code, id) {
				if (id !== "our file id we want to transform") return
				const ast = recast.parse(code)
				// ast transformations
				// This is the ast for the function import
				const functionImportAst = {}
				const astWithImport = await insertAst(ast, functionImportAst, { before: ["body", "0"] })
				const finalAst = await wrapVariableDeclaration(astWithImport, "load", "wrapFn")
				const recastPrint = recast.print()
				// proceed with the transformation...
				return {
					code: recastPrint.code,
					ast: finalAst,
					map: recastPrint.map,
				}
			}, 
		*/
	}
})

export const vitePlugin = unplugin.vite
export const rollupPlugin = unplugin.rollup
