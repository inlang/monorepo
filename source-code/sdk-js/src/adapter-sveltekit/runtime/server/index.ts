export { type SvelteKitServerRuntime, initSvelteKitServerRuntime } from "./runtime.js"
export {
	initHandleWrapper,
	initRootLayoutServerLoadWrapper,
	initServerLoadWrapper,
} from "./wrappers.js"
export {
	initState,
	getResource,
	reloadResources,
	addRuntimeToLocals,
	getRuntimeFromLocals,
} from "./state.js"
