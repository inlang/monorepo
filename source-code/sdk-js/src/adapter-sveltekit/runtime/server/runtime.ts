import type { LanguageTag } from '@inlang/core/languageTag'
import { initRuntimeWithLanguageInformation } from "../../../runtime/index.js"
import { getResource } from "./index.js"

type InitSvelteKitServerRuntimeArgs = {
	sourceLanguageTag: LanguageTag
	languageTags: LanguageTag[]
	languageTag: LanguageTag | undefined
}

export const initSvelteKitServerRuntime = ({
	languageTag,
	sourceLanguageTag,
	languageTags,
}: InitSvelteKitServerRuntimeArgs) => {
	const runtime = initRuntimeWithLanguageInformation({
		readResource: (languageTag: LanguageTag) => getResource(languageTag),
		sourceLanguageTag,
		languageTags,
	})

	if (languageTag) {
		runtime.loadResource(languageTag)
		runtime.changeLanguageTag(languageTag)
	}

	return runtime
}

// TODO: server should also expose a `route` function

export type SvelteKitServerRuntime = Awaited<ReturnType<typeof initSvelteKitServerRuntime>>
