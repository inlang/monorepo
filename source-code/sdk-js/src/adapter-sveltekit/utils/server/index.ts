export { type SvelteKitServerRuntime, initSvelteKitServerRuntime } from "./runtime.js"
export {
	initHandleWrapper,
	initRootLayoutServerLoadWrapper,
	initLayoutServerLoadWrapper,
} from "./wrappers.js"
export {
	referenceLanguage,
	languages,
	getResource,
	addRuntimeToLocals,
	getRuntimeFromLocals,
} from "./state.js"
