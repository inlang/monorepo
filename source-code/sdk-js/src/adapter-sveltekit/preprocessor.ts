import { getConfig } from './config.js'

const REGEX_INLANG_SDK_IMPORT = /.*import\s*{\s*(.*)\s*}\s*from\s+['"]@inlang\/sdk-js['"]/g

const srcFolder = process.cwd() + '/src'

const config = await getConfig()

// ------------------------------------------------------------------------------------------------

type FileType =
	| '+layout.svelte'
	| '+page.svelte'
	| '.svelte'

type FileInformation = {
	type: FileType
	root: boolean
}

const getFileInformation = (id: string): FileInformation | undefined => {
	if (!id.startsWith(srcFolder)) return undefined

	const path = id.replace(srcFolder, '')

	if (path.endsWith('/+layout.svelte')) {
		return {
			type: '+layout.svelte',
			root: path.endsWith('/routes/+layout.svelte'),
		}
	}

	if (path.endsWith('/+page.svelte')) {
		return {
			type: '+page.svelte',
			root: path.endsWith('/routes/+page.svelte'),
		}
	}

	if (path.endsWith('.svelte')) {
		return {
			type: '.svelte',
			root: false,
		}
	}

	return undefined
}

// ------------------------------------------------------------------------------------------------

const transformCode = (code: string, { type, root }: FileInformation) => {
	switch (type) {
		case '+layout.svelte': return transformLayoutSvelte(code, root)
		case '+page.svelte': return transformPageSvelte(code, root)
		case '.svelte': return transformSvelte(code)
	}
}

const transformLayoutSvelte = (code: string, root: boolean) => {
	if (root) return `
<script lang="ts">
	import { getRuntimeFromData } from "@inlang/sdk-js/adapter-sveltekit/shared"
	import { localStorageKey } from "@inlang/sdk-js/adapter-sveltekit/client/reactive"
	import {	getRuntimeFromContext, addRuntimeToContext } from "@inlang/sdk-js/adapter-sveltekit/client/reactive"
	import { browser } from "$app/environment"

	export let data

	addRuntimeToContext(getRuntimeFromData(data))

	let { i, language } = getRuntimeFromContext()

	$: if (browser && $language) {
		document.body.parentElement?.setAttribute("lang", $language)

		localStorage.setItem(localStorageKey, $language)
	}
</script>

{#if $language}
	<slot />
{/if}
`

	return transformSvelte(code)
}

const transformPageSvelte = (code: string, root: boolean) => {
	return transformSvelte(code)
}

// supports multiple imports
// assumption: imports are always on top
const transformSvelte = (code: string): string => {
	let transformedCode: string = code

	let match = REGEX_INLANG_SDK_IMPORT.exec(transformedCode)
	if (!match) return transformedCode

	let lastTransform = 0

	// eslint-disable-next-line no-cond-assign
	while (match) {
		const replacement = (!lastTransform ? 'import { getRuntimeFromContext } from "@inlang/sdk-js/adapter-sveltekit/client/reactive";\n' : '')
			+ 'const {' + match[1] + '} = getRuntimeFromContext();'

		transformedCode = transformedCode.slice(0, match.index)
			+ replacement
			+ transformedCode.slice(match.index + match[0].length)

		lastTransform = match.index + replacement.length
		match = REGEX_INLANG_SDK_IMPORT.exec(transformedCode)
	}

	if (config.isReactive) {
		// replace with store syntax if reactive
		transformedCode = transformedCode.slice(0, lastTransform) + transformedCode.slice(lastTransform)
			.replace(/i\(/g, '$i(')
			.replace(/language[^s]/g, '$language')
	}

	return transformedCode
}

// ------------------------------------------------------------------------------------------------

type PreprocessMarkupArgs = {
	content: string
	filename: string
}

export const preprocessor = {
	async markup({ content, filename }: PreprocessMarkupArgs) {
		const fileInformation = getFileInformation(filename)
		// eslint-disable-next-line unicorn/no-null
		if (!fileInformation) return null

		return { code: transformCode(content, fileInformation) }
	},
}