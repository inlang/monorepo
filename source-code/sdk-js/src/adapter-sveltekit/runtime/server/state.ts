import type { Resource } from "@inlang/core/ast"
import type { InlangConfig, InlangConfigModule } from "@inlang/core/config"
import type { RequestEvent } from "@sveltejs/kit"
import { initConfig } from "../../../config/config.js"
import { inlangSymbol } from "../shared/utils.js"
import type { SvelteKitServerRuntime } from "./runtime.js"
import { languages, referenceLanguage, resources } from "virtual:inlang-static"

let config: InlangConfig

export const initState = async (module: InlangConfigModule) => {
	if (!config) {
		config = import.meta.env.DEV
			? await initConfig(module)
			: ({
					referenceLanguage,
					languages,
					readResources: async () => resources,
					writeResources: async () => undefined,
			  } as InlangConfig)
	}

	await reloadResources()

	return {
		referenceLanguage: config.referenceLanguage,
		languages: config.languages,
	}
}

// ------------------------------------------------------------------------------------------------

let _resources: Resource[] = []

// TODO: fix resources if needed (add missing Keys, etc.)
export const reloadResources = async () => (_resources = await config.readResources({ config }))

export const getResource = (language: string) =>
	_resources.find(({ languageTag: { name } }) => name === language)

// ------------------------------------------------------------------------------------------------

export const addRuntimeToLocals = (
	locals: RequestEvent["locals"],
	runtime: SvelteKitServerRuntime,
) => ((locals as any)[inlangSymbol] = runtime)

export const getRuntimeFromLocals = (locals: RequestEvent["locals"]): SvelteKitServerRuntime =>
	(locals as any)[inlangSymbol]
