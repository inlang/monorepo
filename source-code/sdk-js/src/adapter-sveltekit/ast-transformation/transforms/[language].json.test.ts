import { describe, it, expect } from "vitest"
import { dedent } from "ts-dedent"
import type { TransformConfig } from '../config.js'
import { transformLanguageJson } from "../transforms/[language].json.js"

describe("transformLanguageJson", () => {
	it("adds GET endpoint to an empty file", () => {
		const code = transformLanguageJson({} as TransformConfig, "")
		expect(code).toMatchInlineSnapshot(`
			"
			import { json } from \\"@sveltejs/kit\\"
			import { getResource } from \\"@inlang/sdk-js/adapter-sveltekit/server\\"

			export const GET = (({ params: { language } }) =>
				json(getResource(language) || null))
			"
		`)
	})

	it.skip("adds GET endpoint to a file with arbitrary contents", () => {
		const code = transformLanguageJson({} as TransformConfig, dedent`
			import { error } from "@sveltejs/kit"

			export const GET = () => {
				throw error(500, 'not implemented')
			}
		`)
		expect(code).toMatchInlineSnapshot(`
			"
			import { json, error } from \\"@sveltejs/kit\\"
			import { getResource } from \\"@inlang/sdk-js/adapter-sveltekit/server\\"

			export const GET = (({ params: { language } }) =>
				json(getResource(language) || null))

			export const POST = () => {
				throw Error('not implemented')
			}
			"
		`)
	})

	describe("should throw if GET endpoint is already defined", () => {
		it("arrow function", () => {
			expect(() => transformLanguageJson({} as TransformConfig, dedent`
				import { error } from "@sveltejs/kit"

				export const GET = () => json({ hackerman: true })
			`)).toThrowError()
		})

		it("function keyword", () => {
			expect(() => transformLanguageJson({} as TransformConfig, `
				import { error } from "@sveltejs/kit"

				export async function GET({ params }) {
					return new Response(params.language)
				}
			`)).toThrowError()
		})
	})
})
