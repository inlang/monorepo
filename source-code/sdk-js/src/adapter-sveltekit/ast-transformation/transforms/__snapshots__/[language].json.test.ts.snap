// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformLanguageJson > creates new file 1`] = `
"
import { json } from \\"@sveltejs/kit\\"
import { getResource } from \\"@inlang/sdk-js/adapter-sveltekit/server\\"

export const GET = (({ params: { language } }) =>
	json(getResource(language) || null))
"
`;
